### PR TITLE
Omit BEGIN/COMMIT statements for empty transactions

### LINKED ISSUE #17937: Unnecessary BEGIN/COMMIT happens when model save is a no-op
When a model is saved, and Rails detects that no changes are necessary to send to the server, it still sends a BEGIN/COMMIT to the server.

[Rails console on 4.1.8 showing the issue](https://gist.github.com/Fryguy/2d13e9fafc4a89862d68).  Note that this is using sqlite, but it also happens with PG, which is what we use on our application.

This BEGIN/COMMIT is unnecessary, and contributes to a performance penalty.  The performance penalty is exacerbated by
- having the database on a remote server
- calling save on (thousands) of records where a majority do not change.
  For our application we hit both of those bullet points, and the performance penalty was noticeable.

Unfortunately, when I first discovered this problem our code was running on Rails 2.2.2, a long time ago.  I don't have the numbers on how much performance impact it made anymore, but I remember it being significant for us.  We are currently on Rails 3.2, but I never completed porting the fix to our Rails 3.2 based system, mainly because so much changed in that area of the code.  Now that we are going to Rails 4 and then eventually 5, I am very far removed on how to fix this.  (Not to mention that my personal cycles for fixing are limited).

However, if it helps, I still have the changes I made to fix this for Rails 2.2.2.  The approach was to introduce a delay in the transaction block to not send the "BEGIN" until the first executable statement was also sent.  This worked very well for us, but the down side was every adapter had to be changed to support delaying, which I didn't like.  It felt like there should be a better way to do it for all adapters that support transactions.  This was one reason I never contributed the code to rails proper; I didn't think the fix would be accepted.  

[Delayed BEGIN on Rails 2.2.2 patch](https://gist.github.com/Fryguy/9ea050eda27e5f2ac24a)

EDIT: Clarification on the patch.  This is a patch relative to other patches we had on our code base.  I doubt it's directly applicable to Rails 2.2.2.  It's more meant to give the idea of what we did.

---

@tenderlove This is the stuff we spoke about yesterday.  Hopefully this patch file helps you or whoever might work on it to figure out the best approach.

### PR DESCRIPTION
**Pull request by eugeneius**
Fixes https://github.com/rails/rails/issues/17937.

If a transaction is opened and closed without any queries being run, we can safely omit the `BEGIN` and `COMMIT` statements, as they only exist to modify the connection's behaviour inside the transaction. This removes the overhead of those statements when saving a record with no changes, which makes workarounds like `save if changed?` unnecessary.

~This implementation wraps the raw database connection in a proxy which buffers `begin_transaction` calls, and materializes them the next time the connection is used, or discards them if `end_transaction` is called first. This approach makes new connection usage inside the adapters behave correctly by default. Third party adapters can use the `begin_transaction`/`end_transaction` APIs to opt in to the new behaviour, but will still work even if they don't.~

This implementation buffers transactions inside the transaction manager and materializes them the next time the connection is used. For this to work, the adapter needs to guard all connection use with a call to `materialize_transactions`. Because of this, adapters must opt in to get this new behaviour by implementing `supports_lazy_transactions?`.

If `raw_connection` is used to get a reference to the underlying database connection, the behaviour is disabled and transactions are opened eagerly, as we can't know how the connection will be used. However when the connection is checked back into the pool, we can assume that the application won't use the reference again and reenable lazy transactions. This prevents a single `raw_connection` call from disabling lazy transactions for the lifetime of the connection.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb [hunk 1]
```diff
@@ -112,7 +112,7 @@ def discard! # :nodoc:
       private
 
         def connect
-          @connection = Mysql2::Client.new(@config)
+          @connection = LazyTransactionProxy.new(Mysql2::Client.new(@config))
```
[the remark this answers is not in the document]
Comment (eugeneius) [reply activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: 👍 I've added a private `connection=` method to the abstract adapter that wraps the connection in a proxy, and change the concrete adapters to use it when they create a connection.

**Comment by yahonda** [comment]
@eugeneius Let me take a look at if it "breaks" Oracle enhanced adapter. If it does I am happy to update Oracle enhanced adapter.

**Comment by eugeneius** [comment]
Thanks @yahonda! 🙌 I've tried to make the implementation backwards compatible, so please let me know if it does break - it might mean that I made an incorrect assumption about the adapter interface.

**Comment by eugeneius** [comment]
This test failure was legitimate: https://travis-ci.org/rails/rails/jobs/368924286#L1171-L1175

Executing a prepared statement doesn't touch the connection object (at least on the SQLite adapter), but _does_ require the connection's transaction state to be up to date.

I've fixed the problem by materializing transactions in `exec_query` before prepared statements are executed, but that weakens the "it just works" property that I hoped to get from the proxy approach.

😕

File: activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  module ConnectionAdapters
+    class LazyTransactionProxy < SimpleDelegator # :nodoc:
+      class StatementProxy < SimpleDelegator
```
Comment (sgrif) [inline activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb]: Does it make sense to use `DelegateClass` instead for both of these?

**Review by sgrif — APPROVED** [review]
This feels like a lot of complexity for very little gain, but the implementation seems fine.

File: activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb (hunk 2 above)
Comment (eugeneius) [reply activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb]: I don't think so, because the type of object they wrap depends on the adapter in use; there's no single class we can provide with the full set of methods they need to support.

Comment (eugeneius) [reply activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb]: This comment did make me realise that I forgot to `:nodoc:` the inner class, though.

**Comment by yahonda** [comment]
Oracle enhanced adapter needs some update to support this pull request. Here are CI results for this pull request https://travis-ci.org/yahonda/oracle-enhanced/builds/370051142 . 

I am still having some jet lags due after a long flight, I'll take a look at this result in detail tomorrow.

File: activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+module ActiveRecord
+  module ConnectionAdapters
+    class LazyTransactionProxy < SimpleDelegator # :nodoc:
+      class StatementProxy < SimpleDelegator # :nodoc:
+        def initialize(statement, connection)
+          __setobj__(statement)
+          @connection = connection
+        end
+
+        def method_missing(*)
```
Comment (sgrif) [inline activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb]: This needs to implement `respond_to_missing?` as well

File: activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      def enable_lazy_transactions!
+        @use_lazy_transactions = true
+      end
+
+      def raw_connection
+        disable_lazy_transactions!
+        __getobj__
+      end
+
+      def method_missing(*)
```
Comment (sgrif) [inline activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb]: This needs to implement `respond_to_missing?` as well

File: activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb (hunk 3 above)
Comment (eugeneius) [reply activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb]: `Delegator` already provides an implementation of `respond_to_missing?`:
https://github.com/ruby/ruby/blob/v2_4_4/lib/delegate.rb#L95-L104

Since we call `super` in `method_missing`, we should do the same for `respond_to_missing?`:

```ruby
def respond_to_missing?(*)
  super
end
```

I've added this to both classes, if only to document that using `Delegator`'s behaviour unmodified is intentional.

**Comment by eugeneius** [comment]
@yahonda can you try triggering another build? I renamed `LazyTransactionProxy#raw_connection` to `LazyTransactionProxy#proxied_connection` so it wouldn't clash with [this method](https://github.com/rsim/ruby-plsql/blob/e0b3e01204a8c6c33b3bdbeb2f93f58ba6843a74/lib/plsql/connection.rb#L52).

**Comment by matthewd** [comment]
A few thoughts from an initial read-through:

Yay! :heart: As I mentioned in Pittsburgh, it's awesome to have an implementation to look at -- this one's been sitting on the "nice to have" list for a while.

My gut feeling is that this "should" be closer to the transaction manager rather than the connection -- particularly the deferral stack, though the connection of course needs to be responsible for triggering the materialize.

Failing that, I also wonder about putting the functionality on the abstract connection itself rather than a proxy. I don't mind the _code_ complexity needed for more thorough bookkeeping, but the proxy feels _architecturally_ heavyweight.

Side-thought that would probably just distract us to pursue right now, but I'll mention anyway: this has some interesting parallels (deferring work on the connection until we need to run a real statement, and then potentially giving that some special handling) with #28200 and friends.

Finally, the big one, because it's about behaviour rather than implementation: I believe it's important that savepoints _don't_ materialize the transaction -- that they too get deferred. To me, the ideal version of this change allows us to create transactions more freely, which means we can default `requires_new` to true, eliminating some very surprising behaviours where changes escape a `transaction` block. Reducing a no-op save from two queries to zero is great, of course.. I just see some other shinies in the distance.

**Comment by yahonda** [comment]
CI is green now with Oracle enhanced adapter. As a nature of Oracle database transaction management `begin` and `end` statements are not executed to start and finish transaction. Then some of newly added tests get "failed" but it is okay for me.

**Comment by eugeneius** [comment]
Thanks for the detailed feedback (as always) @matthewd!

I originally wrote a patch that tracked deferred transactions in the transaction manager and materialised them from the adapter, just as you suggested. I wanted to explore the proxy idea instead because I was concerned about the maintenance burden of guarding all connection usage, but if that's an acceptable tradeoff to make for a simpler design, I'll dig up the other implementation and switch to it.

You make a good point that once we have a hook that checks some property of the connection every time it's used and are committed to maintaining that invariant, https://github.com/rails/rails/pull/28200 seems tantalisingly within reach.

I agree that savepoints should be in scope here, as well as "fixture" transactions; the latter is actually what motivated me to work this issue (10 connection pools * 30,000 tests = lots of empty transactions). Adding support for them breaks dozens of tests that assert that an exact queries set of queries are executed though, so I wanted to get feedback on the core of the idea before I fixed them all up.

File: activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb (hunk 3 above)
Comment (eugeneius) [reply activerecord/lib/active_record/connection_adapters/abstract/lazy_transaction_proxy.rb]: The contract between the two methods is that `respond_to_missing?` will return true iff called with a method that `method_missing` would dispatch. Usually this means making equivalent changes to both methods to keep them in sync, as you've noted. In this case however, calling `materialize_transactions` in `method_missing` doesn't affect which methods are dispatched; it's purely a side effect, and so doesn't require an equivalent change in `respond_to_missing?`.

Another way of looking at it: we need to call `materialize_transactions` in `method_missing` because we don't know whether the call to `super` will touch the socket. We don't need to call it in `respond_to_missing?` because we know that calling `super` won't touch the socket.

**Comment by eugeneius** [comment]
I've pushed an alternative implementation that guards all connection use in the adapters instead of wrapping the raw connection in a proxy. The previous implementation is available [here](https://github.com/rails/rails/compare/master...eugeneius:lazy_transactions_proxy).

This version defers savepoints as well as transactions, and is also enabled for "fixture" transactions.

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 5]
```diff
@@ -194,11 +214,41 @@ def begin_transaction(options = {})
...
+
+      def lazy_transactions_enabled?
+        @lazy_transactions_enabled
+      end
+
+      def materialize_transactions
+        return if @materializing_transactions
+        return unless @has_unmaterialized_transactions
+
+        @connection.lock.synchronize do
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: ✅

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 6]
```diff
@@ -132,47 +142,54 @@ def open?; !closed?; end
...
     end
 
     class RealTransaction < Transaction
-      def initialize(connection, options, *args)
-        super
-        if options[:isolation]
-          connection.begin_isolated_db_transaction(options[:isolation])
+      def materialize!
+        if isolation_level
+          connection.begin_isolated_db_transaction(isolation_level)
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: I suspect (at least some?) non-default isolation levels will need to be eagerly materialized... for some, the timing of the BEGIN relative to other connections is relevant, IIRC?

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 7]
```diff
@@ -80,6 +80,8 @@ class AbstractAdapter
       attr_reader :schema_cache, :owner, :logger, :prepared_statements, :lock
       alias :in_use? :owner
 
+      set_callback :checkin, :after, :enable_lazy_transactions!
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]: 🌟

**Review by matthewd — COMMENTED** [review]
😍 

I definitely prefer this approach. Thanks for preserving the previous one for comparison, too.

@sgrif are you more/less/equally inclined toward this version? It spreads out the `materialize_transactions` calls, but to places we already need to care about for other similar reasons (logging, connection locking, future #28200)... that aside, the logic feels more semantically contained to me.

I also think this makes my other future nice-to-have more accessible: if we're three unmaterialized transactions deep when we run a statement, we only _need_ to send a single BEGIN. (If that transaction ends up rolled back, we send a ROLLBACK. If it ends up committed, we send nothing, and mark its parent as having been materialized.) But that's for another time.

**Review by albertoalmagro — APPROVED** [review]
👏 @eugeneius

**Comment by kamipo** [comment]
How much does this change improve the performance of real apps (real use cases)?
I suspect that there is very few empty transactions.

**Comment by matthewd** [comment]
@kamipo I think this is valuable for two use cases: 1) bulk updates, where many/most records aren't actually changed: each `save` is a pair of DB round-trips, with no benefit (but we have to start the transaction before the `before_save`, which means it has to happen before we _know_ there's nothing to write), and 2) nested transactions: we currently have very confusing behaviour around `transaction do` because it defaults to do-nothing if you're already in an outer transaction, but that does surprising things if there's a rollback. I think this change will allow us to default to using proper nested transactions a lot more often, because they become free unless they're used. (And most of the code changes here will be amortized with #28200 -- that needs to hook the same places, so they'll end up sharing the churn cost.)

Basically, I think single empty transactions are likely quite rare, and also uninteresting (a small fixed cost on an unusual request is not much of a concern) -- but there are reasonable-enough patterns that can lead to _lots_ of empty transactions, and while the caller could work around them, it's something we're in a better position to solve.

The two specific use cases I know of are 1) ManageIQ, which carried a custom patch since Rails 2.2 (they lost it during an upgrade, but last I heard it's still an ongoing source of pain), and 2) @jeremy's point in https://github.com/rails/rails/issues/17937#issuecomment-293419214 about not-actual-writes forcing a switch to the primary DB (likely to become a wider complaint as we improve built-in support for such configurations).

**Comment by kamipo** [comment]
I see. Thanks for the clarification, it would help to understand the benefit for future readers (and me).

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 8]
```diff
@@ -203,11 +223,41 @@ def begin_transaction(options = {})
...
+      end
+
+      def materialize_transactions
+        return if @materializing_transactions
+        return unless @has_unmaterialized_transactions
+
+        @connection.lock.synchronize do
+          begin
+            @materializing_transactions = true
+            @stack.drop_while(&:materialized?).each(&:materialize!)
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: Each `drop_while` creates an extra array, so how about adding `return if materialized?` guard in `materialize!`?

Comment (eugeneius) [reply activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: I'm not sure this is worth optimising. The `return unless @has_unmaterialized_transactions` guard already covers the case where there are no transactions to materialize, so we only allocate an extra array when we're about to execute a query, which will result in many allocations.

The use of `drop_while` also expresses an invariant: `@stack` contains zero or more materialized transactions, followed by zero or more unmaterialized transactions. `@stack.each(&:materialize!)` doesn't say anything about the structure of the data.

If you still think this is worth doing, I'd prefer to check `materialized?` inline, both to avoid duplicating the check across both implementations of `materialize!` and to make it clearer that we're not materializing _all_ transactions:

```ruby
@stack.each { |t| t.materialize! unless t.materialized? }
```

Comment (eugeneius) [reply activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: I'm convinced - I pushed an update. Thanks for the explanation @kamipo 🙇

**Comment by matthewd** [comment]
Sorry I left this sitting for ages, and thanks again!

:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by jeremy** [comment]
🚀🌈🚀🌈🚀😊

**Comment by kbrock** [comment]
Thanks so much Matthew and Eugene

**Comment by Nondv** [comment]
Wow. I was just wondering about omiting empty transactions and found out that it was merged several hours ago.

Thank you, rails community. You are awesome <3

**Comment by f3ndot** [comment]
I would love if this would be backported to 5.2. Is that something not out of the realm of possibility?

**Comment by eugeneius** [comment]
I don't think so; 5.2 is only supported for security fixes at this point.

**Comment by pixeltrix** [comment]
@eugeneius I'm pretty sure that 5.2 is still receiving some bug fixes but they'll be limited in scope unless warranted by a new Ruby version or some other similar event (e.g. we backported a whole bunch of stuff to Rails 4.2 to make it work with Ruby 2.4's changes to `to_time`).

However, even given the above I agree that it's unlikely to be backported unless something else makes it necessary.
